### PR TITLE
WIP: Kubernetes service endpoint

### PR DESCRIPTION
**Pull request by mgateva**
## All Submissions:
-------------------------------------
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [YES] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open PRs for the same update/change?

## What is the current behavior?
-------------------------------------

Currently, a service endpoint for Kubernetes cannot be created through this provider, e.g. in a pipeline. 

When you use an Azure Pipeline for an automated deployment to a Kubernetes cluster, you have to create a service endpoint by performing manual actions in the Azure DevOps web-interface.

Whenever you destroy and recreate the Kubernetes cluster, you also have to manually recreate the service endpoint in Azure DevOps.

This breaks automation in such that manual steps are still required prior to deploying objects on a Kubernetes cluster.

Issue number: #221

## What is the new behavior?
-------------------------------------

- This PR adds the functionality to create/destroy a Kubernetes service endpoint.

## Does this introduce a breaking change?
-------------------------------------
- [NO]

## Any relevant logs, error output, etc?
-------------------------------------
n/a

## Other information
-------------------------------------

There are three different authorization method for the Kubernetes service endpoint:
- AzureSubscription
- Kubeconfig
- ServiceAccount

Examples of terraform configuration for each authorization type:

```hcl
resource "azuredevops_serviceendpoint_kubernetes" "serviceendpoint" {
  project_id             = azuredevops_project.project.id
  service_endpoint_name  = "Sample Kubernetes"
  apiserver_url          = "https://sample-kubernetes-cluster.hcp.westeurope.azmk8s.io"
  authorization_type     = "AzureSubscription"
  
  azure_subscription {
    subscription_id   = "8a7aace5-xxxx-xxxx-xxxx-xxxxxxxxxx"
    subscription_name = "Microsoft Azure DEMO"
    tenant_id         = "2e3a33f9-66b1-4xxx-xxxx-xxxxxxxxx"
    resourcegroup_id  = "sample-rg"
    namespace         = "default"
    cluster_name      = "sample-aks"
  }
}

resource "azuredevops_serviceendpoint_kubernetes" "serviceendpoint" {
  project_id             = azuredevops_project.project.id
  service_endpoint_name  = "Sample Kubernetes"
  apiserver_url          = "https://sample-kubernetes-cluster.hcp.westeurope.azmk8s.io"
  authorization_type     = "Kubeconfig"
  
  kubeconfig {
    kube_config = <<EOT
                apiVersion: v1
                clusters:
                - cluster:
                    certificate-authority: fake-ca-file
                    server: https://1.2.3.4
                  name: development
                contexts:
                - context:
                    cluster: development
                    namespace: frontend
                    user: developer
                  name: dev-frontend
                current-context: dev-frontend
                kind: Config
                preferences: {}
                users:
                - name: developer
                  user:
                    client-certificate: fake-cert-file
                    client-key: fake-key-file
                EOT
    accept_untrusted_certs = true
    cluster_context        = "dev-frontend"
  } 
}
                                                  
resource "azuredevops_serviceendpoint_kubernetes" "serviceendpoint" {
  project_id             = azuredevops_project.project.id
  service_endpoint_name  = "Sample Kubernetes"
  apiserver_url          = "https://sample-kubernetes-cluster.hcp.westeurope.azmk8s.io"
  authorization_type     = "ServiceAccount"
  
  service_account {
    token   = "bXktYXBw[...]K8bPxc2uQ=="
    ca_cert = "Mzk1MjgkdmRnN0pi[...]mHHRUH14gw4Q=="
  }
}
```

### CODE REVIEW

**Comment by tmeckel** [comment]
@mgateva I rebased your PR onto the new master ... Feel free to use the branch from here https://github.com/tiwood/terraform-provider-azuredevops/tree/r_serviceendpoint_kubernetes to update your PR

**Comment by tmeckel** [comment]
@mgateva [I just answered @saidst](https://github.com/microsoft/terraform-provider-azuredevops/pull/235#issuecomment-622929668) question about what to do to get your and his PR merged. It would be awesome if you and @saidst could rebase your PRs because the support for Kubernetes Service Endpoints is in high demand. The same with authorizing Service Endpoint connections on projects and pipelines.

**Comment by saidst** [comment]
We are on it. We did most of the work yesterday. Will update this PR asap.
Currently, we work on code redundancy in parts which are not introduced through functionality.

**Comment by nmiodice** [comment]
It looks like https://github.com/microsoft/terraform-provider-azuredevops/pull/235 has the code here + the resource authorization.

Shall we review in https://github.com/microsoft/terraform-provider-azuredevops/pull/235 and close this?

**Comment by mgateva** [comment]
@nmiodice I did some rework on the tests and the flatten function and I was able to run the unit and acc tests successfully. 
Could you please check, why the build is failing?

**Comment by nmiodice** [comment]
Apologies for the day. All day meetings. *sigh*

The failure appeared to be transient so I'm re-running the CI. If it passes, I'll update your branch and we can merge it. I already gave this a thorough review when it was in the other PR.

File: azuredevops/resource_serviceendpoint_kubernetes_test.go [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+
+	var attrTestCheckFuncList []resource.TestCheckFunc
+	attrTestCheckFuncList = append(
+		attrTestCheckFuncList,
+		resource.TestCheckResourceAttrSet(terraformServiceEndpointNode, "kubeconfig.#"),
+	)
+	testAccAzureDevOpsServiceEndpoint(t, attrTestCheckFuncList, authorizationType)
+}
+
+func testAccAzureDevOpsServiceEndpoint(t *testing.T, attrTestCheckFuncList []resource.TestCheckFunc, authorizationType string) {
```
Comment (nmiodice) [inline azuredevops/resource_serviceendpoint_kubernetes_test.go:516]: I really like this approach to reduce the code needed to write acceptance tests!

**Comment by nmiodice** [comment]
Pending successful CI, this LGTM!

**Pull request merged**